kornia/augmentation/container/utils.py

```diff
@@ -0,0 +1,367 @@
+from abc import ABCMeta, abstractmethod
+from functools import partial
+from typing import Callable, cast, Dict, Iterator, List, Optional, Tuple, Type, Union
+
+import torch
+import torch.nn as nn
+
+import kornia  # lazy loading for circular dependencies
+from kornia.augmentation.base import (
+    _AugmentationBase,
+    GeometricAugmentationBase2D,
+    MixAugmentationBase,
+    TensorWithTransformMat,
+)
+from kornia.constants import DataKey
+from kornia.geometry.bbox import transform_bbox
+from kornia.geometry.linalg import transform_points
+from kornia.utils.helpers import _torch_inverse_cast
+
+from .base import ParamItem
+
+
+def _get_geometric_only_param(
+    module: "kornia.augmentation.container.ImageSequential", param: List[ParamItem]
+) -> List[ParamItem]:
+    named_modules: Iterator[Tuple[str, nn.Module]] = module.get_forward_sequence(param)
+
+    res: List[ParamItem] = []
+    for (_, mod), p in zip(named_modules, param):
+        if isinstance(mod, (GeometricAugmentationBase2D,)):
+            res.append(p)
+    return res
+
+
+class ApplyInverseInterface(metaclass=ABCMeta):
+    """Abstract interface for applying and inversing transformations."""
+
+    @classmethod
+    @abstractmethod
+    def apply_trans(
+        cls,
+        input: torch.Tensor,
+        label: Optional[torch.Tensor],
+        module: nn.Module,
+        param: ParamItem,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """Apply a transformation with respect to the parameters.
+
+        Args:
+            input: the input tensor.
+            label: the optional label tensor.
+            module: any torch Module but only kornia augmentation modules will count
+                to apply transformations.
+            param: the corresponding parameters to the module.
+        """
+        pass
```

Comment:
Abstract method does not raise `NotImplementedError`
https://deepsource.io/gh/kornia/kornia/run/3492ca94-6546-4722-8a23-f570db6834d1/python/PTC-W0053

same for the other methods